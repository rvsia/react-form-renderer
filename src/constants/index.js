export const components = {
  TEXT_FIELD: 'text-field',
  TEXTAREA_FIELD: 'textarea-field',
  FIELD_ARRAY: 'field-array',
  SELECT_COMPONENT: 'select-field',
  FIXED_LIST: 'fixed-list',
  CHECKBOX: 'checkbox',
  SUB_FORM: 'sub-form',
  RADIO: 'radio',
  TABS: 'tabs',
  TAB_ITEM: 'tab-item',
  DATE_PICKER: 'date-picker',
  TIME_PICKER: 'time-picker',
  TAG_CONTROL: 'tag-control',
  WIZARD: 'wizard',
};
  /** Validator functions placeholder */
export const validators = {
  REQUIRED: 'required-validator',
  /**
     * min length of the input value
     */
  MIN_LENGTH: 'min-length-validator',
  /**
     * max length of the input value
     */
  MAX_LENGTH: 'max-length-validator',
  /**
   * exact length of input value
   */
  EXACT_LENGTH: 'exact-length-validator',
  /**
     * minimum count of fileds in some dynamic list of fields
     */
  MIN_ITEMS_VALIDATOR: 'min-items-validator',
  /**
     * Minimum value of number input
     */
  MIN_NUMBER_VALUE: 'min-number-value',
  /**
     * Maximum value of number inpuy
     */
  MAX_NUMBER_VALUE: 'max-number-value',
  /**
     * Regexp pattern validator
     */
  PATTERN_VALIDATOR: 'pattern-validator',
};

export const layoutComponents = {
  FORM_WRAPPER: 'FormWrapper',
  BUTTON: 'Button',
  COL: 'Col',
  FORM_GROUP: 'FormGroup',
  BUTTON_GROUP: 'ButtonGroup',
  ICON: 'Icon',
  ARRAY_FIELD_WRAPPER: 'ArrayFieldWrapper',
  HELP_BLOCK: 'HelpBlock',
};
