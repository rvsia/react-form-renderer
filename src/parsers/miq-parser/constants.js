import { components } from '../../constants';

export const componentMap = {
  DialogFieldTextBox: components.TEXT_FIELD,
  DialogFieldRadioButton: components.RADIO,
  DialogFieldCheckBox: components.CHECKBOX,
  DialogFieldTextAreaBox: components.TEXTAREA_FIELD,
  DialogFieldDropDownList: components.SELECT_COMPONENT,
  DialogFieldDateControl: components.DATE_PICKER,
  DialogFieldDateTimeControl: components.TIME_PICKER,
  DialogFieldTagControl: components.TAG_CONTROL,
};

// for true values
export const neededAttributes = [
  'name',
  'label',
  [ 'label', 'title' ],
  [ 'data_type', 'dataType' ],
  [ 'required', 'isRequired' ],
  [ 'read_only', 'isReadOnly' ],
  [ 'description', 'helperText' ],
];

// for false values
export const neededFalseAttributes = [
  [ 'visible', 'isVisible' ],
];
