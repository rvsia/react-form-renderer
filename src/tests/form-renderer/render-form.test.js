import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Form } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import renderForm from '../../form-renderer/render-form';
import RendererContext, { configureContext } from '../../form-renderer/renderer-context';
import { components, validators } from '../../constants';

describe('renderForm function', () => {
  let layoutMapper;

  const ContextWrapper = ({ children, ...props }) => (
    <RendererContext.Provider value={ configureContext({
      ...props,
    }) }>
      <Form onSubmit={ jest.fn() } mutators={{ ...arrayMutators }}>
        { () =>  children }
      </Form>
    </RendererContext.Provider>
  );

  beforeEach(() => {
    layoutMapper = {
      Col: ({ children }) => <div>{ children }</div>,
      FormGroup: ({ children }) => <div>{ children }</div>,
      ButtonGroup: ({ children }) => <div>{ children }</div>,
      HelpBlock: ({ children }) => <div>{ children }</div>,
      Button: ({ label, bsStyle, ...rest }) => <button { ...rest }>{ label }</button>,
      Icon: ({ type, name }) => <div>Icon: { name }</div>,
    };
  });

  it('should render single field from defined componentTypes', () => {
    const formFields = [{
      component: components.TEXT_FIELD,
      name: 'foo',
    }];
    const wrapper = mount(
      <ContextWrapper formFieldsMapper={{
        [components.TEXT_FIELD]: ({ FieldProvider, dataType, ...props }) => <div { ...props }>TextField</div>,
      }}>
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should correctly assign dataType validator if no additional validators given', () => {
    const formFields = [{
      component: components.TEXT_FIELD,
      name: 'foo',
      dataType: 'string',
    }];
    const wrapper = mount(
      <ContextWrapper formFieldsMapper={{
        [components.TEXT_FIELD]: ({ FieldProvider, dataType, ...props }) => <div { ...props }>TextField</div>,
      }}>
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );
    const form = wrapper.find(Form);
    form.instance().form.change('foo', 1);
    expect(form.instance().state.state.errors.foo).toBe(('Field value has to be string'));
  });

  it('should correctly assign required validator with custom message', () => {
    const formFields = [{
      component: components.TEXT_FIELD,
      name: 'foo',
      dataType: 'string',
      validate: [{
        type: validators.REQUIRED,
        message: 'Bar',
      }],
    }];
    const wrapper = mount(
      <ContextWrapper formFieldsMapper={{
        [components.TEXT_FIELD]: ({ FieldProvider, dataType, ...props }) => <div { ...props }>TextField</div>,
      }}>
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );
    const form = wrapper.find(Form);
    expect(form.instance().state.state.errors.foo).toBe(('Bar'));
  });

  it('should render single field from with custom componentType', () => {
    const formFields = [{
      component: 'custom-component',
      name: 'foo',
    }];
    const wrapper = mount(
      <ContextWrapper formFieldsMapper={{
        'custom-component': ({ FieldProvider, dataType, formOptions, ...props }) => <div { ...props }>Custom component</div>,
      }}>
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render single field from with custom componentType and assign FieldProvider', () => {
    const formFields = [{
      component: 'custom-component',
      name: 'foo',
      assignFieldProvider: true,
    }];

    const wrapper = mount(
      <ContextWrapper formFieldsMapper={{
        'custom-component': ({ FieldProvider, dataType, formOptions, ...props }) => (
          <FieldProvider { ...props } render={ () => <div>Custom component</div> } />
        ),
      }}>
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render array field', () => {
    const formFields = [{
      component: components.FIELD_ARRAY,
      name: 'foo',
      key: 'bar',
      fields: [],
    }];

    const wrapper = mount(
      <ContextWrapper layoutMapper={ layoutMapper }>
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should correctly assign array field validators', () => {
    const formFields = [{
      component: components.FIELD_ARRAY,
      name: 'foo',
      key: 'bar',
      validate: [{
        type: validators.REQUIRED,
      }, {
        type: validators.MIN_ITEMS_VALIDATOR,
        treshold: 3,
      }],
      fields: [{
        component: components.TEXT_FIELD,
        name: 'name',
        type: 'text',
      }],
    }];

    const wrapper = mount(
      <ContextWrapper
        layoutMapper={ layoutMapper }
        formFieldsMapper={{
          [components.TEXT_FIELD]: ({ FieldProvider, dataType, ...props }) => <div { ...props }>TextField</div>,
        }}
      >
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );

    const form = wrapper.find(Form);
    form.instance().form.mutators.push('foo');
    expect(form.instance().state.state.errors.foo).toBeTruthy();
  });

  it('should render condition field only if the condition is met', () => {
    const formFields = [{
      component: 'custom-component',
      name: 'foo',
      condition: {
        when: 'bar',
        is: 'fuzz',
      },
    }];

    const wrapper = mount(
      <ContextWrapper formFieldsMapper={{
        'custom-component': ({ FieldProvider, dataType, formOptions, ...props }) => <div { ...props }>Custom component</div>,
      }}>
        { renderForm(formFields, { renderForm }) }
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find(Form).instance().form.change('bar', 'fuzz');
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
