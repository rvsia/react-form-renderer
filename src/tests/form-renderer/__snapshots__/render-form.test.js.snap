// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderForm function should render array field 1`] = `
<ContextWrapper
  layoutMapper={
    Object {
      "Button": [Function],
      "ButtonGroup": [Function],
      "Col": [Function],
      "FormGroup": [Function],
      "HelpBlock": [Function],
      "Icon": [Function],
    }
  }
>
  <ReactFinalForm
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper>
      <FieldWrapper
        component={[Function]}
        componentType="field-array"
        fields={Array []}
        formOptions={
          Object {
            "renderForm": [Function],
          }
        }
        key="bar"
        name="foo"
        validate={
          Array [
            undefined,
          ]
        }
      >
        <_class
          FieldProvider={[Function]}
          arrayValidator={[Function]}
          component={[Function]}
          fields={Array []}
          formOptions={
            Object {
              "hasFixedItems": false,
              "renderForm": [Function],
            }
          }
          name="foo"
        >
          <Field
            FieldProvider={[Function]}
            arrayValidator={[Function]}
            component={[Function]}
            fields={Array []}
            formOptions={
              Object {
                "hasFixedItems": false,
                "renderForm": [Function],
              }
            }
            format={[Function]}
            name="foo"
            parse={[Function]}
            reactFinalForm={
              Object {
                "batch": [Function],
                "blur": [Function],
                "change": [Function],
                "focus": [Function],
                "getFieldState": [Function],
                "getRegisteredFields": [Function],
                "getState": [Function],
                "initialize": [Function],
                "isValidationPaused": [Function],
                "mutators": Object {
                  "insert": [Function],
                  "move": [Function],
                  "pop": [Function],
                  "push": [Function],
                  "remove": [Function],
                  "shift": [Function],
                  "swap": [Function],
                  "unshift": [Function],
                  "update": [Function],
                },
                "pauseValidation": [Function],
                "registerField": [Function],
                "reset": [Function],
                "resumeValidation": [Function],
                "setConfig": [Function],
                "submit": [Function],
                "subscribe": [Function],
              }
            }
          >
            <Component
              FieldProvider={[Function]}
              arrayValidator={[Function]}
              fields={Array []}
              formOptions={
                Object {
                  "hasFixedItems": false,
                  "renderForm": [Function],
                }
              }
              input={
                Object {
                  "name": "foo",
                  "onBlur": [Function],
                  "onChange": [Function],
                  "onFocus": [Function],
                  "value": "",
                }
              }
              meta={
                Object {
                  "active": false,
                  "data": Object {},
                  "dirty": false,
                  "dirtySinceLastSubmit": false,
                  "error": undefined,
                  "initial": undefined,
                  "invalid": false,
                  "pristine": true,
                  "submitError": undefined,
                  "submitFailed": false,
                  "submitSucceeded": false,
                  "submitting": false,
                  "touched": false,
                  "valid": true,
                  "visited": false,
                }
              }
            >
              <renderArrayField
                FieldProvider={[Function]}
                arrayValidator={[Function]}
                fields={Array []}
                formOptions={
                  Object {
                    "hasFixedItems": false,
                    "renderForm": [Function],
                  }
                }
                input={
                  Object {
                    "name": "foo",
                    "onBlur": [Function],
                    "onChange": [Function],
                    "onFocus": [Function],
                    "value": "",
                  }
                }
                meta={
                  Object {
                    "active": false,
                    "data": Object {},
                    "dirty": false,
                    "dirtySinceLastSubmit": false,
                    "error": undefined,
                    "initial": undefined,
                    "invalid": false,
                    "pristine": true,
                    "submitError": undefined,
                    "submitFailed": false,
                    "submitSucceeded": false,
                    "submitting": false,
                    "touched": false,
                    "valid": true,
                    "visited": false,
                  }
                }
                validate={Array []}
              >
                <DynamicArray
                  FieldProvider={[Function]}
                  arrayValidator={[Function]}
                  fieldKey="foo"
                  fields={Array []}
                  formOptions={
                    Object {
                      "hasFixedItems": false,
                      "renderForm": [Function],
                    }
                  }
                  input={
                    Object {
                      "name": "foo",
                      "onBlur": [Function],
                      "onChange": [Function],
                      "onFocus": [Function],
                      "value": "",
                    }
                  }
                  meta={
                    Object {
                      "active": false,
                      "data": Object {},
                      "dirty": false,
                      "dirtySinceLastSubmit": false,
                      "error": undefined,
                      "initial": undefined,
                      "invalid": false,
                      "pristine": true,
                      "submitError": undefined,
                      "submitFailed": false,
                      "submitSucceeded": false,
                      "submitting": false,
                      "touched": false,
                      "valid": true,
                      "visited": false,
                    }
                  }
                  renderForm={[Function]}
                  validate={Array []}
                >
                  <_class
                    key="foo"
                    name="foo"
                    validate={[Function]}
                  >
                    <ReactFinalFormFieldArray(4.8.1)(2.0.1)
                      name="foo"
                      reactFinalForm={
                        Object {
                          "batch": [Function],
                          "blur": [Function],
                          "change": [Function],
                          "focus": [Function],
                          "getFieldState": [Function],
                          "getRegisteredFields": [Function],
                          "getState": [Function],
                          "initialize": [Function],
                          "isValidationPaused": [Function],
                          "mutators": Object {
                            "insert": [Function],
                            "move": [Function],
                            "pop": [Function],
                            "push": [Function],
                            "remove": [Function],
                            "shift": [Function],
                            "swap": [Function],
                            "unshift": [Function],
                            "update": [Function],
                          },
                          "pauseValidation": [Function],
                          "registerField": [Function],
                          "reset": [Function],
                          "resumeValidation": [Function],
                          "setConfig": [Function],
                          "submit": [Function],
                          "subscribe": [Function],
                        }
                      }
                      validate={[Function]}
                    >
                      <Col
                        xs={11}
                      >
                        <div />
                      </Col>
                      <Col
                        className="final-form-array-add-container"
                        xs={1}
                      >
                        <div>
                          <FormGroup>
                            <div>
                              <ButtonGroup
                                className="pull-right"
                              >
                                <div>
                                  <Button
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    <button
                                      onClick={[Function]}
                                      type="button"
                                    />
                                  </Button>
                                </div>
                              </ButtonGroup>
                            </div>
                          </FormGroup>
                        </div>
                      </Col>
                    </ReactFinalFormFieldArray(4.8.1)(2.0.1)>
                  </_class>
                </DynamicArray>
              </renderArrayField>
            </Component>
          </Field>
        </_class>
      </FieldWrapper>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`renderForm function should render condition field only if the condition is met 1`] = `
<ContextWrapper
  formFieldsMapper={
    Object {
      "custom-component": [Function],
    }
  }
>
  <ReactFinalForm
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper
      condition={
        Object {
          "is": "fuzz",
          "when": "bar",
        }
      }
    >
      <Condition
        is="fuzz"
        when="bar"
      >
        <_class
          name="bar"
          subscription={
            Object {
              "value": true,
            }
          }
        >
          <Field
            format={[Function]}
            name="bar"
            parse={[Function]}
            reactFinalForm={
              Object {
                "batch": [Function],
                "blur": [Function],
                "change": [Function],
                "focus": [Function],
                "getFieldState": [Function],
                "getRegisteredFields": [Function],
                "getState": [Function],
                "initialize": [Function],
                "isValidationPaused": [Function],
                "mutators": Object {
                  "insert": [Function],
                  "move": [Function],
                  "pop": [Function],
                  "push": [Function],
                  "remove": [Function],
                  "shift": [Function],
                  "swap": [Function],
                  "unshift": [Function],
                  "update": [Function],
                },
                "pauseValidation": [Function],
                "registerField": [Function],
                "reset": [Function],
                "resumeValidation": [Function],
                "setConfig": [Function],
                "submit": [Function],
                "subscribe": [Function],
              }
            }
            subscription={
              Object {
                "value": true,
              }
            }
          />
        </_class>
      </Condition>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`renderForm function should render condition field only if the condition is met 2`] = `
<ContextWrapper
  formFieldsMapper={
    Object {
      "custom-component": [Function],
    }
  }
>
  <ReactFinalForm
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper
      condition={
        Object {
          "is": "fuzz",
          "when": "bar",
        }
      }
    >
      <Condition
        is="fuzz"
        when="bar"
      >
        <_class
          name="bar"
          subscription={
            Object {
              "value": true,
            }
          }
        >
          <Field
            format={[Function]}
            name="bar"
            parse={[Function]}
            reactFinalForm={
              Object {
                "batch": [Function],
                "blur": [Function],
                "change": [Function],
                "focus": [Function],
                "getFieldState": [Function],
                "getRegisteredFields": [Function],
                "getState": [Function],
                "initialize": [Function],
                "isValidationPaused": [Function],
                "mutators": Object {
                  "insert": [Function],
                  "move": [Function],
                  "pop": [Function],
                  "push": [Function],
                  "remove": [Function],
                  "shift": [Function],
                  "swap": [Function],
                  "unshift": [Function],
                  "update": [Function],
                },
                "pauseValidation": [Function],
                "registerField": [Function],
                "reset": [Function],
                "resumeValidation": [Function],
                "setConfig": [Function],
                "submit": [Function],
                "subscribe": [Function],
              }
            }
            subscription={
              Object {
                "value": true,
              }
            }
          >
            <FieldWrapper
              component={[Function]}
              componentType="custom-component"
              formOptions={
                Object {
                  "renderForm": [Function],
                }
              }
              name="foo"
              validate={
                Array [
                  undefined,
                ]
              }
            >
              <customComponent
                FieldProvider={[Function]}
                formOptions={
                  Object {
                    "renderForm": [Function],
                  }
                }
                name="foo"
                validate={[Function]}
              >
                <div
                  name="foo"
                  validate={[Function]}
                >
                  Custom component
                </div>
              </customComponent>
            </FieldWrapper>
          </Field>
        </_class>
      </Condition>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`renderForm function should render single field form with custom componentType and assign FieldProvider 1`] = `
<ContextWrapper
  formFieldsMapper={
    Object {
      "custom-component": [Function],
    }
  }
>
  <ReactFinalForm
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper>
      <FieldWrapper
        assignFieldProvider={true}
        component={[Function]}
        componentType="custom-component"
        formOptions={
          Object {
            "renderForm": [Function],
          }
        }
        name="foo"
        validate={
          Array [
            undefined,
          ]
        }
      >
        <customComponent
          FieldProvider={[Function]}
          formOptions={
            Object {
              "renderForm": [Function],
            }
          }
          name="foo"
          validate={[Function]}
        >
          <_class
            name="foo"
            render={[Function]}
            validate={[Function]}
          >
            <Field
              format={[Function]}
              name="foo"
              parse={[Function]}
              reactFinalForm={
                Object {
                  "batch": [Function],
                  "blur": [Function],
                  "change": [Function],
                  "focus": [Function],
                  "getFieldState": [Function],
                  "getRegisteredFields": [Function],
                  "getState": [Function],
                  "initialize": [Function],
                  "isValidationPaused": [Function],
                  "mutators": Object {
                    "insert": [Function],
                    "move": [Function],
                    "pop": [Function],
                    "push": [Function],
                    "remove": [Function],
                    "shift": [Function],
                    "swap": [Function],
                    "unshift": [Function],
                    "update": [Function],
                  },
                  "pauseValidation": [Function],
                  "registerField": [Function],
                  "reset": [Function],
                  "resumeValidation": [Function],
                  "setConfig": [Function],
                  "submit": [Function],
                  "subscribe": [Function],
                }
              }
              render={[Function]}
              validate={[Function]}
            >
              <div>
                Custom component
              </div>
            </Field>
          </_class>
        </customComponent>
      </FieldWrapper>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`renderForm function should render single field from defined componentTypes 1`] = `
<ContextWrapper
  formFieldsMapper={
    Object {
      "text-field": [Function],
    }
  }
>
  <ReactFinalForm
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper>
      <FieldWrapper
        component={[Function]}
        componentType="text-field"
        formOptions={
          Object {
            "renderForm": [Function],
          }
        }
        name="foo"
        validate={
          Array [
            undefined,
          ]
        }
      >
        <_class
          FieldProvider={[Function]}
          component={[Function]}
          formOptions={
            Object {
              "renderForm": [Function],
            }
          }
          name="foo"
          validate={[Function]}
        >
          <Field
            FieldProvider={[Function]}
            component={[Function]}
            formOptions={
              Object {
                "renderForm": [Function],
              }
            }
            format={[Function]}
            name="foo"
            parse={[Function]}
            reactFinalForm={
              Object {
                "batch": [Function],
                "blur": [Function],
                "change": [Function],
                "focus": [Function],
                "getFieldState": [Function],
                "getRegisteredFields": [Function],
                "getState": [Function],
                "initialize": [Function],
                "isValidationPaused": [Function],
                "mutators": Object {
                  "insert": [Function],
                  "move": [Function],
                  "pop": [Function],
                  "push": [Function],
                  "remove": [Function],
                  "shift": [Function],
                  "swap": [Function],
                  "unshift": [Function],
                  "update": [Function],
                },
                "pauseValidation": [Function],
                "registerField": [Function],
                "reset": [Function],
                "resumeValidation": [Function],
                "setConfig": [Function],
                "submit": [Function],
                "subscribe": [Function],
              }
            }
            validate={[Function]}
          >
            <Component
              FieldProvider={[Function]}
              formOptions={
                Object {
                  "renderForm": [Function],
                }
              }
              input={
                Object {
                  "name": "foo",
                  "onBlur": [Function],
                  "onChange": [Function],
                  "onFocus": [Function],
                  "value": "",
                }
              }
              meta={
                Object {
                  "active": false,
                  "data": Object {},
                  "dirty": false,
                  "dirtySinceLastSubmit": false,
                  "error": undefined,
                  "initial": undefined,
                  "invalid": false,
                  "pristine": true,
                  "submitError": undefined,
                  "submitFailed": false,
                  "submitSucceeded": false,
                  "submitting": false,
                  "touched": false,
                  "valid": true,
                  "visited": false,
                }
              }
            >
              <div
                formOptions={
                  Object {
                    "renderForm": [Function],
                  }
                }
                input={
                  Object {
                    "name": "foo",
                    "onBlur": [Function],
                    "onChange": [Function],
                    "onFocus": [Function],
                    "value": "",
                  }
                }
                meta={
                  Object {
                    "active": false,
                    "data": Object {},
                    "dirty": false,
                    "dirtySinceLastSubmit": false,
                    "error": undefined,
                    "initial": undefined,
                    "invalid": false,
                    "pristine": true,
                    "submitError": undefined,
                    "submitFailed": false,
                    "submitSucceeded": false,
                    "submitting": false,
                    "touched": false,
                    "valid": true,
                    "visited": false,
                  }
                }
              >
                TextField
              </div>
            </Component>
          </Field>
        </_class>
      </FieldWrapper>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`renderForm function should render single field from with custom componentType 1`] = `
<ContextWrapper
  formFieldsMapper={
    Object {
      "custom-component": [Function],
    }
  }
>
  <ReactFinalForm
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper>
      <FieldWrapper
        component={[Function]}
        componentType="custom-component"
        formOptions={
          Object {
            "renderForm": [Function],
          }
        }
        name="foo"
        validate={
          Array [
            undefined,
          ]
        }
      >
        <customComponent
          FieldProvider={[Function]}
          formOptions={
            Object {
              "renderForm": [Function],
            }
          }
          name="foo"
          validate={[Function]}
        >
          <div
            name="foo"
            validate={[Function]}
          >
            Custom component
          </div>
        </customComponent>
      </FieldWrapper>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;
